lib/using_restricted_array.rb

```diff
@@ -6,58 +6,114 @@
+    # The run time will depend on how many elements are in the array
+# Space complexity: O(1)  The space needed to store the count doesn't depend on the array size,
+#as the count value gets overridden until we reach the final count
 def length(array)
-  raise NotImplementedError
+  count = 0
+  x = 0
+  until array[x] == nil
+    count += 1
+    x += 1
+  end
+  return count
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)N represents the number of elements in the array.
+    # The run time will depend on how many elements are in the array
+# Space complexity: O(n) N reperesents each output from the array. 
+#The space needed for storage depends on how many elements are outputed from the array.
 def print_array(array)
-  raise NotImplementedError
+  x = 0
+  until array[x] == nil
+    puts array[x]
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) N represents the number of elements in the array.
+    # The run time will depend on how many elements are in the array
+# Space complexity: O(1) The program will return true or fall no matter the size of the array.
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  x = 0
+  until array[x] == nil
+    return true if array[x] == value_to_find
+    x += 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) N represents the number of elements in the array.
+    # The run time will depend on how many elements are in the array
+# Space complexity: O(1). The program will only return the 
+#largest value no matter the size of the array
 def find_largest(array, length)
-  raise NotImplementedError
+  largest = array[0]
+  x = 0
+  until array[x] == nil
+    largest = array[x] if array[x] > largest
+    x += 1
+  end
+  return largest
+  
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1). The program will only return the 
+#smallest value no matter the size of the array
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest = array[0]
+  x = 0
+  until array[x] == nil
+    smallest = array[x] if array[x] < smallest
+    x += 1
+  end
+  return smallest
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)N represents the number of elements in the array.
+    # The run time will depend on how many elements are in the array
+# Space complexity: O(n) The space needed for the reverse array depends on the size of the original array.
 def reverse(array, length)
```

Comment:
The trick here is you're not reversing the original array, but returning a copy which is reversed.  The tests are expecting the original array to be reversed.